We will focus on 
importing datasets, 
dealing with missing values, 
changing data types, 
filtering, sorting, 
selecting specific column(s), 
dealing with duplicate values, 
dropping and adding rows and columns, 
counting values, 
counting unique values.

In [1]:
# import libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

C:\Users\PARI\anaconda3\lib\site-packages\scipy\__init__.py:138: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.24.4)
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion} is required for this version of "


In [9]:
# import dataset
df = pd.read_csv("train.csv")

In [10]:
# Checking first elements
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [11]:
# Checking last elements
df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

### Reading a CSV file with specific columns

Let’s assume that you don’t want all of the columns in your CSV file. You just want 3 of them and you want to get rid of them at the begining. How to do that? Good news! There is a really easy way to do that. We are going to use usecols argument to specify the column names that we want to work with. Let’s work with just PassengerId, Survived, Pclass columns.

In [12]:
df1 = pd.read_csv('train.csv', usecols=['PassengerId','Survived','Pclass'])

In [13]:
df1.head()

PassengerId  Survived  Pclass
0            1         0       3
1            2         1       1
2            3         1       3
3            4         1       1
4            5         0       3

### Getting some basic information about data

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [15]:
df.shape

(891, 12)

In [16]:
df.size

10692

In [18]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [24]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [23]:
df.describe(include='all')

PassengerId    Survived      Pclass                     Name   Sex  \
count    891.000000  891.000000  891.000000                      891   891   
unique          NaN         NaN         NaN                      891     2   
top             NaN         NaN         NaN  Braund, Mr. Owen Harris  male   
freq            NaN         NaN         NaN                        1   577   
mean     446.000000    0.383838    2.308642                      NaN   NaN   
std      257.353842    0.486592    0.836071                      NaN   NaN   
min        1.000000    0.000000    1.000000                      NaN   NaN   
25%      223.500000    0.000000    2.000000                      NaN   NaN   
50%      446.000000    0.000000    3.000000                      NaN   NaN   
75%      668.500000    1.000000    3.000000                      NaN   NaN   
max      891.000000    1.000000    3.000000                      NaN   NaN   

               Age       SibSp       Parch  Ticket        Fare    Cabin  \
count   714.000000  891.000000  891.000000     891  891.000000      204   
unique         NaN         NaN         NaN     681         NaN      147   
top            NaN         NaN         NaN  347082         NaN  B96 B98   
freq           NaN         NaN         NaN       7         NaN        4   
mean     29.699118    0.523008    0.381594     NaN   32.204208      NaN   
std      14.526497    1.102743    0.806057     NaN   49.693429      NaN   
min       0.420000    0.000000    0.000000     NaN    0.000000      NaN   
25%      20.125000    0.000000    0.000000     NaN    7.910400      NaN   
50%      28.000000    0.000000    0.000000     NaN   14.454200      NaN   
75%      38.000000    1.000000    0.000000     NaN   31.000000      NaN   
max      80.000000    8.000000    6.000000     NaN  512.329200      NaN   

       Embarked  
count       889  
unique        3  
top           S  
freq        644  
mean        NaN  
std         NaN  
min         NaN  
25%         NaN  
50%         NaN  
75%         NaN  
max         NaN

We can see that count of Age column is 714, mean is 29.6, standard deviation is 14.52 and so on. Thanks to count, we can understand that there are some missing values in this column. We will deal with them later.

### Sorting the DataFrame based on a column
Sorting numeric data

In [20]:
df.sort_values('Fare').head(10)

PassengerId  Survived  Pclass                             Name   Sex  \
271          272         1       3     Tornquist, Mr. William Henry  male   
597          598         0       3              Johnson, Mr. Alfred  male   
302          303         0       3  Johnson, Mr. William Cahoone Jr  male   
633          634         0       1    Parr, Mr. William Henry Marsh  male   
277          278         0       2      Parkes, Mr. Francis "Frank"  male   
413          414         0       2   Cunningham, Mr. Alfred Fleming  male   
674          675         0       2       Watson, Mr. Ennis Hastings  male   
263          264         0       1            Harrison, Mr. William  male   
466          467         0       2            Campbell, Mr. William  male   
732          733         0       2             Knight, Mr. Robert J  male   

      Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
271  25.0      0      0    LINE   0.0   NaN        S  
597  49.0      0      0    LINE   0.0   NaN        S  
302  19.0      0      0    LINE   0.0   NaN        S  
633   NaN      0      0  112052   0.0   NaN        S  
277   NaN      0      0  239853   0.0   NaN        S  
413   NaN      0      0  239853   0.0   NaN        S  
674   NaN      0      0  239856   0.0   NaN        S  
263  40.0      0      0  112059   0.0   B94        S  
466   NaN      0      0  239853   0.0   NaN        S  
732   NaN      0      0  239855   0.0   NaN        S

In [21]:
df.sort_values('Fare',ascending=False)

PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
737          738         1       1              Lesurer, Mr. Gustave J   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
88            89         1       1          Fortune, Miss. Mabel Helen   
27            28         0       1      Fortune, Mr. Charles Alexander   
..           ...       ...     ...                                 ...   
633          634         0       1       Parr, Mr. William Henry Marsh   
413          414         0       2      Cunningham, Mr. Alfred Fleming   
822          823         0       1     Reuchlin, Jonkheer. John George   
732          733         0       2                Knight, Mr. Robert J   
674          675         0       2          Watson, Mr. Ennis Hastings   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  
258  female  35.0      0      0  PC 17755  512.3292          NaN        C  
737    male  35.0      0      0  PC 17755  512.3292         B101        C  
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C  
88   female  23.0      3      2     19950  263.0000  C23 C25 C27        S  
27     male  19.0      3      2     19950  263.0000  C23 C25 C27        S  
..      ...   ...    ...    ...       ...       ...          ...      ...  
633    male   NaN      0      0    112052    0.0000          NaN        S  
413    male   NaN      0      0    239853    0.0000          NaN        S  
822    male  38.0      0      0     19972    0.0000          NaN        S  
732    male   NaN      0      0    239855    0.0000          NaN        S  
674    male   NaN      0      0    239856    0.0000          NaN        S  

[891 rows x 12 columns]

#### Sorting columns with strings

In [22]:
df['Cabin'].describe()

count         204
unique        147
top       B96 B98
freq            4
Name: Cabin, dtype: object

In [26]:
df.sort_values('Cabin').head(10)

PassengerId  Survived  Pclass  \
583          584         0       1   
475          476         0       1   
556          557         1       1   
284          285         0       1   
599          600         1       1   
630          631         1       1   
867          868         0       1   
647          648         1       1   
209          210         1       1   
185          186         0       1   

                                                  Name     Sex   Age  SibSp  \
583                                Ross, Mr. John Hugo    male  36.0      0   
475                        Clifford, Mr. George Quincy    male   NaN      0   
556  Duff Gordon, Lady. (Lucille Christiana Sutherl...  female  48.0      1   
284                         Smith, Mr. Richard William    male   NaN      0   
599       Duff Gordon, Sir. Cosmo Edmund ("Mr Morgan")    male  49.0      1   
630               Barkworth, Mr. Algernon Henry Wilson    male  80.0      0   
867               Roebling, Mr. Washington Augustus II    male  31.0      0   
647                Simonius-Blumer, Col. Oberst Alfons    male  56.0      0   
209                                   Blank, Mr. Henry    male  40.0      0   
185                              Rood, Mr. Hugh Roscoe    male   NaN      0   

     Parch    Ticket     Fare Cabin Embarked  
583      0     13049  40.1250   A10        C  
475      0    110465  52.0000   A14        S  
556      0     11755  39.6000   A16        C  
284      0    113056  26.0000   A19        S  
599      0  PC 17485  56.9292   A20        C  
630      0     27042  30.0000   A23        S  
867      0  PC 17590  50.4958   A24        S  
647      0     13213  35.5000   A26        C  
209      0    112277  31.0000   A31        C  
185      0    113767  50.0000   A32        S

In [28]:
df.sort_values('Cabin', ascending=False).head(10)

PassengerId  Survived  Pclass  \
339          340         0       1   
394          395         1       3   
205          206         0       3   
251          252         0       3   
10            11         1       3   
183          184         1       2   
618          619         1       2   
776          777         0       3   
66            67         1       2   
345          346         1       2   

                                                  Name     Sex   Age  SibSp  \
339                       Blackwell, Mr. Stephen Weart    male  45.0      0   
394  Sandstrom, Mrs. Hjalmar (Agnes Charlotta Bengt...  female  24.0      0   
205                         Strom, Miss. Telma Matilda  female   2.0      0   
251         Strom, Mrs. Wilhelm (Elna Matilda Persson)  female  29.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
183                          Becker, Master. Richard F    male   1.0      2   
618                        Becker, Miss. Marion Louise  female   4.0      2   
776                                   Tobin, Mr. Roger    male   NaN      0   
66                        Nye, Mrs. (Elizabeth Ramell)  female  29.0      0   
345                      Brown, Miss. Amelia "Mildred"  female  24.0      0   

     Parch      Ticket     Fare Cabin Embarked  
339      0      113784  35.5000     T        S  
394      2     PP 9549  16.7000    G6        S  
205      1      347054  10.4625    G6        S  
251      1      347054  10.4625    G6        S  
10       1     PP 9549  16.7000    G6        S  
183      1      230136  39.0000    F4        S  
618      1      230136  39.0000    F4        S  
776      0      383121   7.7500   F38        Q  
66       0  C.A. 29395  10.5000   F33        S  
345      0      248733  13.0000   F33        S

There are lots of missing (NaN) values in this column. How can we deal with them? Thankfully, there is an argument which is called na_position which helps us to set a position for the NaN values in the dataset.

In [32]:
df.sort_values('Cabin',ascending=True, na_position='last')

PassengerId  Survived  Pclass  \
583          584         0       1   
475          476         0       1   
556          557         1       1   
284          285         0       1   
599          600         1       1   
..           ...       ...     ...   
884          885         0       3   
885          886         0       3   
886          887         0       2   
888          889         0       3   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
583                                Ross, Mr. John Hugo    male  36.0      0   
475                        Clifford, Mr. George Quincy    male   NaN      0   
556  Duff Gordon, Lady. (Lucille Christiana Sutherl...  female  48.0      1   
284                         Smith, Mr. Richard William    male   NaN      0   
599       Duff Gordon, Sir. Cosmo Edmund ("Mr Morgan")    male  49.0      1   
..                                                 ...     ...   ...    ...   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch           Ticket     Fare Cabin Embarked  
583      0            13049  40.1250   A10        C  
475      0           110465  52.0000   A14        S  
556      0            11755  39.6000   A16        C  
284      0           113056  26.0000   A19        S  
599      0         PC 17485  56.9292   A20        C  
..     ...              ...      ...   ...      ...  
884      0  SOTON/OQ 392076   7.0500   NaN        S  
885      5           382652  29.1250   NaN        Q  
886      0           211536  13.0000   NaN        S  
888      2       W./C. 6607  23.4500   NaN        S  
890      0           370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [33]:
df.sort_values('Cabin',ascending=False, na_position='last')

PassengerId  Survived  Pclass  \
339          340         0       1   
394          395         1       3   
205          206         0       3   
251          252         0       3   
10            11         1       3   
..           ...       ...     ...   
884          885         0       3   
885          886         0       3   
886          887         0       2   
888          889         0       3   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
339                       Blackwell, Mr. Stephen Weart    male  45.0      0   
394  Sandstrom, Mrs. Hjalmar (Agnes Charlotta Bengt...  female  24.0      0   
205                         Strom, Miss. Telma Matilda  female   2.0      0   
251         Strom, Mrs. Wilhelm (Elna Matilda Persson)  female  29.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
..                                                 ...     ...   ...    ...   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch           Ticket     Fare Cabin Embarked  
339      0           113784  35.5000     T        S  
394      2          PP 9549  16.7000    G6        S  
205      1           347054  10.4625    G6        S  
251      1           347054  10.4625    G6        S  
10       1          PP 9549  16.7000    G6        S  
..     ...              ...      ...   ...      ...  
884      0  SOTON/OQ 392076   7.0500   NaN        S  
885      5           382652  29.1250   NaN        Q  
886      0           211536  13.0000   NaN        S  
888      2       W./C. 6607  23.4500   NaN        S  
890      0           370376   7.7500   NaN        Q  

[891 rows x 12 columns]

### Counting the occurences of variables

In [35]:
df.Sex

0        male
1      female
2      female
3      female
4        male
        ...  
886      male
887    female
888    female
889      male
890      male
Name: Sex, Length: 891, dtype: object

In [39]:
df.Sex.value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

In [44]:
df.nunique()
#by typing this, we can see the counts of unique numbers in each column

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

In [45]:
df['Sex'].unique()

array(['male', 'female'], dtype=object)

In [46]:
df['Sex'].nunique()

2

In [51]:
df[['Embarked','Sex']].nunique()

Embarked    3
Sex         2
dtype: int64

#### Changing the data type

In [52]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [53]:
df['Embarked'] = df['Embarked'].astype('category')

In [54]:
df.Embarked.dtypes

CategoricalDtype(categories=['C', 'Q', 'S'], ordered=False)

In [55]:
df.dtypes

PassengerId       int64
Survived          int64
Pclass            int64
Name             object
Sex              object
Age             float64
SibSp             int64
Parch             int64
Ticket           object
Fare            float64
Cabin            object
Embarked       category
dtype: object

## Filtering

#### Filtering under one condition

In [56]:
df["Embarked"] == "C" 

0      False
1       True
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889     True
890    False
Name: Embarked, Length: 891, dtype: bool

In [58]:
#If we write it like that, pandas will understand that we want to see those rows that have True value
df[df['Embarked']=='C']

PassengerId  Survived  Pclass  \
1              2         1       1   
9             10         1       2   
19            20         1       3   
26            27         0       3   
30            31         0       1   
..           ...       ...     ...   
866          867         1       2   
874          875         1       2   
875          876         1       3   
879          880         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN      0   
26                             Emir, Mr. Farred Chehab    male   NaN      0   
30                            Uruchurtu, Don. Manuel E    male  40.0      0   
..                                                 ...     ...   ...    ...   
866                       Duran y More, Miss. Asuncion  female  27.0      1   
874              Abelson, Mrs. Samuel (Hannah Wizosky)  female  28.0      1   
875                   Najib, Miss. Adele Kiamie "Jane"  female  15.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch         Ticket     Fare Cabin Embarked  
1        0       PC 17599  71.2833   C85        C  
9        0         237736  30.0708   NaN        C  
19       0           2649   7.2250   NaN        C  
26       0           2631   7.2250   NaN        C  
30       0       PC 17601  27.7208   NaN        C  
..     ...            ...      ...   ...      ...  
866      0  SC/PARIS 2149  13.8583   NaN        C  
874      0      P/PP 3381  24.0000   NaN        C  
875      0           2667   7.2250   NaN        C  
879      1          11767  83.1583   C50        C  
889      0         111369  30.0000  C148        C  

[168 rows x 12 columns]

#### Filtering under two or more condition

We are going to use AND and OR operator to filter with more than one condition. Let’s assume that we want to see the passengers whose Fare is smaller than 100 and who are female. We are going to create 2 new masks to complete that.

In [60]:
df_fare_mask = df["Fare"] < 100
df_sex_mask = df["Sex"] == "female"
df[df_fare_mask & df_sex_mask]

PassengerId  Survived  Pclass  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
8              9         1       3   
9             10         1       2   
..           ...       ...     ...   
880          881         1       2   
882          883         0       3   
885          886         0       3   
887          888         1       1   
888          889         0       3   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
882                       Dahlberg, Miss. Gerda Ulrika  female  22.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   

     Parch            Ticket     Fare Cabin Embarked  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
8        2            347742  11.1333   NaN        S  
9        0            237736  30.0708   NaN        C  
..     ...               ...      ...   ...      ...  
880      1            230433  26.0000   NaN        S  
882      0              7552  10.5167   NaN        S  
885      5            382652  29.1250   NaN        Q  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  

[280 rows x 12 columns]

In [62]:
df_fare_mask2 = df["Fare"] > 500
df_age_mask = df["Age"] > 70
df[df_fare_mask2 | df_age_mask]

PassengerId  Survived  Pclass                                  Name  \
96            97         0       1             Goldschmidt, Mr. George B   
116          117         0       3                  Connors, Mr. Patrick   
258          259         1       1                      Ward, Miss. Anna   
493          494         0       1               Artagaveytia, Mr. Ramon   
630          631         1       1  Barkworth, Mr. Algernon Henry Wilson   
679          680         1       1    Cardeza, Mr. Thomas Drake Martinez   
737          738         1       1                Lesurer, Mr. Gustave J   
851          852         0       3                   Svensson, Mr. Johan   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  
96     male  71.0      0      0  PC 17754   34.6542           A5        C  
116    male  70.5      0      0    370369    7.7500          NaN        Q  
258  female  35.0      0      0  PC 17755  512.3292          NaN        C  
493    male  71.0      0      0  PC 17609   49.5042          NaN        C  
630    male  80.0      0      0     27042   30.0000          A23        S  
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C  
737    male  35.0      0      0  PC 17755  512.3292         B101        C  
851    male  74.0      0      0    347060    7.7750          NaN        S

### Finding the null values with .isnull()

In [63]:
df.isnull()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
886        False     False   False  False  False  False  False  False   False   
887        False     False   False  False  False  False  False  False   False   
888        False     False   False  False  False   True  False  False   False   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked  
0    False   True     False  
1    False  False     False  
2    False   True     False  
3    False  False     False  
4    False   True     False  
..     ...    ...       ...  
886  False   True     False  
887  False  False     False  
888  False   True     False  
889  False  False     False  
890  False   True     False  

[891 rows x 12 columns]

In [64]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### Dealing with missing values

In [65]:
df2 = df.copy()

#### Dropping a column

In [66]:
df2.drop(labels = ["Cabin"], axis=1).head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

In [69]:
df2.drop(labels = ["Cabin", "Name"], axis=1).head()

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch  \
0            1         0       3    male  22.0      1      0   
1            2         1       1  female  38.0      1      0   
2            3         1       3  female  26.0      0      0   
3            4         1       1  female  35.0      1      0   
4            5         0       3    male  35.0      0      0   

             Ticket     Fare Embarked  
0         A/5 21171   7.2500        S  
1          PC 17599  71.2833        C  
2  STON/O2. 3101282   7.9250        S  
3            113803  53.1000        S  
4            373450   8.0500        S

#### Filling missing values with .fillna()

In [72]:
#df2["columnname"].fillna(0, inplace = True) 

In [73]:
df['Age'] = df['Age'].fillna((df['Age'].median()))